falcon/util/streams.py

```diff
@@ -0,0 +1,276 @@
+                if delimiter_pos >= 0:
+                    return self._finalize_read_until(
+                        amount, result, have_bytes,
+                        delimiter_pos=delimiter_pos)
+
+            if amount < (have_bytes + self._buffer_len - self._buffer_pos -
+                         delimiter_len_1):
+                # NOTE(vytas) We now have enough data in the buffer to
+                # return to the caller.
+                return self._finalize_read_until(amount, result, have_bytes)
+
+            read_amount = self._chunk_size
+            if read_amount > self._max_bytes_remaining:
+                read_amount = self._max_bytes_remaining
+            if read_amount == 0:
+                return self._finalize_read_until(
+                    amount, result, have_bytes, delimiter, delimiter_pos,
+                    missing_delimiter_error=missing_delimiter_error)
+
+            next_chunk = self._read_func(read_amount)
+            next_chunk_len = len(next_chunk)
+            self._max_bytes_remaining -= read_amount
+            if next_chunk_len < self._chunk_size:
+                self._buffer_len += next_chunk_len
+                self._buffer += next_chunk
+                return self._finalize_read_until(
+                    amount, result, have_bytes, delimiter,
+                    missing_delimiter_error=missing_delimiter_error)
+
+            # NOTE(vytas) The buffer was empty before, skip straight to the
+            #   next chunk.
+            if self._buffer_len <= self._buffer_pos:
+                self._buffer_len = next_chunk_len
+                self._buffer_pos = 0
+                self._buffer = next_chunk
+                continue
+
+            if delimiter_len_1 > 0:
+                offset = max(self._buffer_len - delimiter_len_1,
+                             self._buffer_pos)
+                fragment = (self._buffer[offset:] +
+                            next_chunk[:delimiter_len_1])
+                delimiter_pos = fragment.find(delimiter)
+                if delimiter_pos >= 0:
+                    self._buffer_len += next_chunk_len
+                    self._buffer += next_chunk
+                    return self._finalize_read_until(
+                        amount, result, have_bytes, delimiter,
+                        delimiter_pos + offset)
+
+            if have_bytes + self._buffer_len - self._buffer_pos >= amount:
+                # NOTE(vytas): we have now verified that all bytes currently in
+                #   the buffer are delimiter-free, including the border of the
+                #   upcoming chunk
+                return self._finalize_read_until(
+                    amount, result, have_bytes, next_chunk=next_chunk,
+                    next_chunk_len=next_chunk_len)
+
+            have_bytes += self._buffer_len - self._buffer_pos
+            if self._buffer_pos > 0:
+                result.append(self._buffer[self._buffer_pos:])
```

Comment:
I did a few quick benchmarks (CPython 3.8) using variations of the test provided in one of the answers to [this SO question](https://stackoverflow.com/questions/9099145/where-are-python-bytearrays-used). The `b''.join()` pattern is indeed more performant. When combining four 8K chunks the difference on my box was something like 1.14 µs vs. 2.43 µs. The difference seemed to increase non-linearly vs. the number of chunks.

On the other hand, `bytearray.extend()` has the advantage of letting you free the chunk immediately rather than building up a list of them and holding them all in memory. This briefly requires double the memory (after the `b''.join()` but before the list of chunks is deallocated).

Reply:
I suppose we could just make sure the docs make it clear that people should avoid reading an entire stream in one call (vs. piping it to the destination) since it could eat up a lot of memory. But some people will still ignore that or not read the docs closely. So I wonder if we could do something adaptive? Read the first few chunks in to the list, but at some threshold switch over to to extending a `bytearray`?

Reply:
I was benchmarking a bit and `io.BytesIO` surprisingly seems to do the right thing™ bringing the best of both worlds, so I'll probably incorporate it into the solution one or another way.

Reply:
Added a shim using `BytesIO`.